cql2pgjson/src/main/java/org/folio/cql2pgjson/CQL2PgJSON.java

```diff
@@ -919,25 +925,66 @@ private String queryByLike(String index, DbIndex dbIndex, IndexTextAndJsonValues
    * @return
    */
   private String queryBySql(DbIndex dbIndex, IndexTextAndJsonValues vals, CQLTermNode node, String comparator, CqlModifiers modifiers) {
-    String index = vals.getIndexText();
-
+    String indexMod = vals.getIndexText();
     if (comparator.equals("==")) {
       comparator = "=";
     }
+    Index schemaIndex = dbIndex.getIndex();
+    String sql;
     String term = "'" + Cql2SqlUtil.cql2like(node.getTerm()) + "'";
     if (CqlTermFormat.NUMBER.equals(modifiers.getCqlTermFormat())) {
-      index = "(" + index + ")::numeric";
-      term = node.getTerm();
+      sql = "(" + indexMod + ")::numeric " + comparator + term;
+    } else if(schemaIndex != null) {
+      sql = createSQLLengthCase(comparator, indexMod, term, false,schemaIndex);
+    } else {
+      sql = indexMod + " " + comparator + term;
     }
-    String sql = index + " " + comparator + term;
 
     if (! dbIndex.hasIndex() && ! dbIndex.hasFullTextIndex() && ! dbIndex.hasLikeIndex()) {
-      String s = String.format("%s, CQL >>> SQL: %s >>> %s", index, node.toCQL(), sql);
+      String s = String.format("%s, CQL >>> SQL: %s >>> %s", indexMod, node.toCQL(), sql);
       logger.log(Level.WARNING, "Doing SQL query without index for {0}", s);
     }
 
-    logger.log(Level.FINE, "index {0} generated SQL {1}", new Object[] {index, sql});
+    logger.log(Level.FINE, "index {0} generated SQL {1}", new Object[] {indexMod, sql});
+    return sql;
+  }
+
+  private String createSQLLengthCase(String comparator, String index, String term, boolean not, Index schemaIndex) {
+    String sql;
+    sql = "CASE WHEN length(" + wrapInLowerUnaccent(term, schemaIndex) + ") <= 600 THEN " + wrapInLowerUnaccent(index, schemaIndex) + " " + comparator + " " + wrapInLowerUnaccent(term, schemaIndex);
+    String lengthCaseComparator = determineLengthCaseComparator(comparator);
+    sql = appendLengthCase(lengthCaseComparator,comparator, wrapInLowerUnaccent(index, schemaIndex), wrapInLowerUnaccent(term, schemaIndex), not, sql);
+    return sql;
```

Comment:
This method is difficult to read and understand.
"CASE WHEN" and "THEN" is here, but "ELSE ... END" is hidden in the method appendLengthCase.
Can you move all the code from appendLengthCase into this method?
Then it is possible to directly return the result without a temporary sql variable:

    return "CASE WHEN length(" + wrapInLowerUnaccent(term, schemaIndex) + ...

(This is also faster because the compiler automatically uses a StringBuilder for string concatenation).

Can you limit the line length to 120 characters and use line breaks and indentation for readability? Example: https://github.com/folio-org/raml-module-builder/blob/v29.1.2/domain-models-runtime/src/main/java/org/folio/rest/persist/PgUtil.java#L1021